### PR TITLE
Calculate min and max ssz lengths

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
In preparation for validating rpc message lengths more strictly, calculate the minimum and maximum lengths for SSZ serialised objects.  Not currently used anywhere but changes to `SimpleOffsetSerializer` are so complex it's worth reviewing them separately.

## Fixed Issue(s)
Part of #2077 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        return Bytes4.SIZE;
+      case "BLSSignature":
+        return BLSSignature.BLS_SIGNATURE_SIZE;
+      case "BLSPublicKey":
+        return BLSPublicKey.BLS_PUBKEY_SIZE;
+      case "Boolean":
+      case "boolean":
+        return BOOLEAN_SIZE;
+      default:
+        throw new IllegalArgumentException("Unable to deserialize " + classInfo.getSimpleName());
```
Comment (Nashatyrev) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java]: May be change exception wording?

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    fieldLengthBounds =
+        new LengthBounds(
+            0, elementLengthBounds.getMax() * listElementMaxSize + variableFieldOffsetsLength);
+    return fieldLengthBounds;
+  }
+
+  private static LengthBounds calculateBitlistLength(
+      final ReflectionInformation reflectionInfo, final int variableFieldCount) {
+    final LengthBounds fieldLengthBounds;
+    final long maxSize = reflectionInfo.getBitlistElementMaxSizes().get(variableFieldCount);
```
Comment (Nashatyrev) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java:123]: Not sure here: `getBitlistElementMaxSizes()` seems to return _only_ `Bitlist` instance sizes in the referred class. Why is it indexed by `variableFieldCount` which is a total counter of lists and bitfields?

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            vectorLength * elementLengthBounds.getMin(),
+            vectorLength * elementLengthBounds.getMax());
+    return fieldLengthBounds;
+  }
+
+  private static LengthBounds calculateSszListLength(
+      final ReflectionInformation reflectionInfo, final int variableFieldCount) {
+    final LengthBounds fieldLengthBounds;
+    final Class<?> listElementType = reflectionInfo.getListElementTypes().get(variableFieldCount);
+    final long listElementMaxSize = reflectionInfo.getListElementMaxSizes().get(variableFieldCount);
```
Comment (Nashatyrev) [inline ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java:110]: The same question as for `calculateBitlistLength`

**Review by Nashatyrev — APPROVED** [review]
Looks good 👍  Just make sure that `variableFieldCount` is used correctly

TBH overall SSZ code looks not too good :) Another chance to think on SSZ refactoring.

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java]: Done.

File: ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java (hunk 2 above)
Comment (ajsutton) [reply ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/util/LengthBoundCalculator.java:123]: That is *highly* suspicious but exactly matches what `SimpleOffsetSerializer` does. It also works for every type we know how to SSZ serialise, but I strongly suspect that's a coincidence and there just aren't any objects that have both `SSZList` and `Bitlist`.

Created a test class with both and sure enough it fails with `IndexOutOfBoundsException`.  So have fixed both `LengthBoundsCalculator` and `SimpleOffsetSerializer`.

**Pull request merged**